tests/test_connector.py

```diff
@@ -2064,6 +2064,39 @@ async def handler(request):
     await conn.close()
 
 
+async def test_tcp_connector_fqdn(
+    aiohttp_server: Any, ssl_ctx: Any, client_ssl_ctx: Any
+) -> None:
+    async def handler(request):
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+
+    srv = await aiohttp_server(app, ssl=ssl_ctx)
+    port = unused_port()
+    conn = aiohttp.TCPConnector(local_addr=("127.0.0.1", port))
+
+    session = aiohttp.ClientSession(connector=conn)
+    url = srv.make_url("/")
+
+    r = await session.get(url.with_host("localhost."), ssl=client_ssl_ctx)
```

Comment:
Parametrized the `test_tcp_connector_do_not_raise_connector_ssl_error` test.

* Can't add `example.org` because it does not resolve to local host, so the connection fails.
* Can't add `localhost...` because the resolution fails during IDNA parsing (that processes each component separately).

I don't think there is any value in supporting multiple dots at the end of the host name, so I did not change the IDNA code. Please let me know your thoughts about this.

Reply:
> I don't think there is any value in supporting multiple dots at the end of the host name, so I did not change the IDNA code. Please let me know your thoughts about this.

I think this is valuable in terms of consistency with what other popular projects do, so long as it doesn't violate RFCs.

@martin-sucha could you fix this too? I think we should be consistent across the entire code base, rather than supporting this only in some places. Where/how does this fail exactly?

Reply:
> so long as it doesn't violate RFCs.

I can't say it does not as I haven't verified it. Personally I prefer the more strict/conservative handling of not allowing multiple dots at the end.

> could you fix this too? I think we should be consistent across the entire code base, rather than supporting this only in some places.

Okay, added the stripping to `_create_dirrect_connection`.

> Where/how does this fail exactly?

```
../aiohttp/connector.py:856: in _resolve_host
    addrs = await self._resolver.resolve(host, port, family=self._family)
        host       = 'localhost...'
        key        = ('localhost...', 35905)
        port       = 35905
        self       = <aiohttp.connector.TCPConnector object at 0x7f542fc5c790>
        traces     = []
```

raises

```
UnicodeError: encoding with 'idna' codec failed (UnicodeError: label empty or too long)
```

Reply:
@martin-sucha thanks! Let's fix that IDNA thing too, for consistency.

Reply:
Looks like this is the last request that still needs to be implemented.

Reply:
This one is already implemented by the replacement of multiple trailing dots with a single dot in `_create_direct_connection` that we discussed in https://github.com/aio-libs/aiohttp/pull/7364#discussion_r1299426737

The UnicodeError will not be raised because there won't be empty labels in the DNS name.

Reply:
@martin-sucha sounds like we need a test for that, then? Maybe with some actual Unicode chars? Both decoded and encoded with punnycode?

Reply:
I don't understand why we'd need another test.

`test_tcp_connector_do_not_raise_connector_ssl_error`/`fully-qualified domain name with multiple trailing dots` covers the case.

We just won't pass a string with multiple trailing dots to `self._resolve_host` so we don't need to make any changes in `_resolve_host`.

Reply:
I was just trying to cover the cases that might happen if someone introduces a but in the future. You implied that there may be problems with IDN in some places so I thought it's a good idea to have regression tests covering such corner cases and expecting the same across the board. But if you think it's too much, I won't insist.

Reply:
Marking this thread as resolved.